cirq/google/calibration/phased_fsim.py

```diff
@@ -130,3 +140,217 @@ def override_by(self, other: 'PhasedFSimCharacterization') -> 'PhasedFSimCharact
             other that are not None). Otherwise the current values are used.
         """
         return other.merge_with(self)
+
+
+# TODO: Add start and end calibration timestamp
+# TODO: Add export to Panda's data frame
+@dataclasses.dataclass(frozen=True)
+class PhasedFSimCalibrationResult:
+    """The PhasedFSimGate characterization result.
+
+    Attributes:
+        parameters: Map from qubit pair to characterization result. For each pair of characterized
+            quibts a and b either only (a, b) or only (b, a) is present.
+        gate: Characterized gate for each qubit pair.
+    """
+
+    # TODO: Add validation that only either (a, b) or (b, a) is present.
+    parameters: Dict[Tuple[Qid, Qid], PhasedFSimCharacterization]
+    gate: Gate
```

Comment:
why do we need `gate` as well? Isn't it a `cirq.PhasedFSimGate` always? otherwise the parameters type signature wouldn't make sense

Reply:
I put Gate becuase it can be Union[ISwapPowGate, PhasedFSimGate, FSimGate, PhasedISwapPowGate], anything which can serialize to a compatible gate.

I guess we could make a decision to support only PhasedFSimGate here but I guess this is a broader question on what is the strategy when using gates in Cirq?

Reply:
But what is it actually used for? Is it possible to have `PhasedFSimCharacterization` with 5 values but `gate` is ISwapPowGate? Also this is a continuation of the comment below where I think `gate` in the `..Result` object is just to preserve the input context? It doesn't actually change the results, right?

Reply:
Now I see what you mean. `Gate` is anything which is compatible and handled by the `SerializableGateSet` used for the engine execution. Since the `SerializableGateSet` accept many forms of a gate, I copied this logic here. This is reminiscence of the logic adopted in Cirq where we don't really have a notion of a "hardware gate", instead we say for example `criq.CZ` and this means: run the approximate `cirq.CZ` gate on this qubit pair. If we have a notion of hardware gate which is not fixed to any specific parameters but is just an instruction to do, that might be maybe less confusing. However, it is also questionable.

Answering your questions:
 - It is perfectly fine to have all 5 parameters in `PhasedFSimCharacterization` for `ISwapPowGate` because `ISwapPowGate` is just an approximation to the perfect gate, and `PhasedFSimCharacterization` tells you distance to this approximation in this particular context.
 -  `gate` in `...Result` is just for context, right.

At some point options will not only have requests to characterize but also to calibrate, in which case some of the angles will be automatically fixed to perfect values, and in the end `PhasedFSimCharacterization` will be a perfect gate that was requested for.

Reply:
I take back one statement above: in the PR that I'm doing right now I'll be using gate parameter to verify that the circuit being compiled is compatible with characterization.

Reply:
Good point about the lack of the "hardware gate" notion. I would have thought that the `PhasedFSim...` family of classes could *implicitly* specify the hardware gate (specifically "PhasedFSimGate"), but I guess there's a couple of hardware gates (sqrtiswap, syc, iswap, ...) that all fit under that umbrella so we need to request a particular "ideal" gate.

Reply:
Even `FSimGate` covers all the hardware gates we support. The problem is that depending how people create the circuits, they can be in many forms. We don't have a notion of optimizer/compiler which could normalize the gate used so I need to keep all the options in this API.